### PR TITLE
First PR

### PR DESCRIPTION
**Pull request by zmitic**

### CODE REVIEW

File: src/Fixture/SimpleProductFixture.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusRestockNotificationPlugin\Fixture;
+
+use Doctrine\Common\Persistence\ObjectManager;
+use Sylius\Bundle\FixturesBundle\Fixture\AbstractFixture;
+
+class SimpleProductFixture extends AbstractFixture
```
Comment (loevgaard) [inline src/Fixture/SimpleProductFixture.php]: Wrong name

File: src/Fixture/SimpleProductFixture.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    private $objectManager;
+
+    public function __construct(ObjectManager $objectManager)
+    {
+        $this->objectManager = $objectManager;
+    }
+
+    public function load(array $options): void
+    {
+        dd($this->objectManager);
```
Comment (loevgaard) [inline src/Fixture/SimpleProductFixture.php]: :D

File: src/Form/Extension/ProductVariantMatchExtension.php [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+        $view->vars['attr']['data-notify'] = $url;
+    }
+
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+    }
+
+    /** @noinspection OverridingDeprecatedMethodInspection */
+    public function getExtendedType(): string
```
Comment (loevgaard) [inline src/Form/Extension/ProductVariantMatchExtension.php]: Remove this method

File: src/Form/Extension/ProductVariantMatchExtension.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+use Symfony\Component\Form\AbstractTypeExtension;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\Form\FormView;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Modify error message when variant option is not in stock
+ */
+class ProductVariantMatchExtension extends AbstractTypeExtension
```
Comment (loevgaard) [inline src/Form/Extension/ProductVariantMatchExtension.php]: If you extend the `ProductOptionValueChoiceType`, the name of class should be `ProductOptionValueChoiceTypeExtension`

File: src/Form/Validator/CartItemAvailability.php [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Setono\SyliusRestockNotificationPlugin\Form\Validator;
+
+class CartItemAvailability
```
Comment (loevgaard) [inline src/Form/Validator/CartItemAvailability.php]: What is this?

File: src/Listener/ProductRestockedListener.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        if (!$entity instanceof ProductVariantInterface) {
+            return;
+        }
+
+        $changeSet = $event->getEntityChangeSet();
+        if (!isset($changeSet['onHand'])) {
+            return;
+        }
+
+//        $previous = $changeSet['onHand'][0];
```
Comment (loevgaard) [inline src/Listener/ProductRestockedListener.php]: Shouldn't this code be commented back in?

You should somehow check that the product isn't already in stock

File: src/Listener/ProductRestockedListener.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+//        if ($previous > 0) {
+//            return;
+//        }
+
+        $product = $entity->getProduct();
+        if (!$product) {
+            return;
+        }
+
+        $id = $product->getId();
```
Comment (loevgaard) [inline src/Listener/ProductRestockedListener.php]:
```suggestion
```

File: src/Listener/ProductRestockedListener.php [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        $command = new NotifyCustomersOnRestockCommand($id);
```
Comment (loevgaard) [inline src/Listener/ProductRestockedListener.php]:
```suggestion
        $command = new NotifyCustomersOnRestockCommand($product->getId());
```

File: src/Listener/ProductRestockedListener.php [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            return;
+        }
+
+//        $previous = $changeSet['onHand'][0];
+//        if ($previous > 0) {
+//            return;
+//        }
+
+        $product = $entity->getProduct();
+        if (!$product) {
```
Comment (loevgaard) [inline src/Listener/ProductRestockedListener.php]:
```suggestion
        if (null === $product) {
```

File: src/Listener/ProductRestockedListener.php (hunk 6 above)
Comment (loevgaard) [reply src/Listener/ProductRestockedListener.php]: You also have to check if the `onHand` value is `> 0`

File: src/Entity/StockNotification.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+class StockNotification implements ResourceInterface, TimestampableInterface
+{
+    use TimestampableTrait;
+
+    private $id;
+    private $product;
+    private $email;
+    private $locale;
+
+    public function __construct(ProductInterface $product, string $email, string $locale)
```
Comment (loevgaard) [inline src/Entity/StockNotification.php]: It should be a `ProductVariantInterface` instance and not a product here. A `ProductInterface` doesn't have stock, a `ProductVariantInterface` does

File: src/Resources/config/config.yml [hunk 11]
```diff
@@ -0,0 +1,23 @@
+imports:
+    - { resource: services.yml }
+
+framework:
+    messenger:
+        buses:
+            setono_restock_notifications.command_bus: ~
+
+sylius_resource:
```
Comment (loevgaard) [inline src/Resources/config/config.yml]: See the guidelines for resources: https://github.com/Setono/Coding-Guidelines#5-resources

File: src/Resources/config/config.yml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    resources:
+        setono_sylius_restock_notification.notification:
+            classes:
+                model: Setono\SyliusRestockNotificationPlugin\Entity\StockNotification
+#                interface: Setono\SyliusRestockNotificationPlugin\Entity\StockNotificationInterface
```
Comment (loevgaard) [inline src/Resources/config/config.yml]:
```suggestion
```

File: src/Resources/config/config.yml [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+                repository: Setono\SyliusRestockNotificationPlugin\Repository\StockNotificationRepository
+#                form: BitBag\SyliusInvoicingPlugin\Form\Type\InvoiceType
```
Comment (loevgaard) [inline src/Resources/config/config.yml]:
```suggestion
```

File: src/Resources/config/config.yml [hunk 14]
```diff
@@ -0,0 +1,23 @@
+imports:
```
Comment (loevgaard) [inline src/Resources/config/config.yml]: All Symfony yaml files should have `.yaml` extension

File: src/Resources/config/doctrine/StockNotification.orm.xml [hunk 15]
```diff
@@ -0,0 +1,34 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<doctrine-mapping xmlns="http://doctrine-project.org/schemas/orm/doctrine-mapping"
+                  xmlns:gedmo="http://gediminasm.org/schemas/orm/doctrine-extensions-mapping"
+                  xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                  xsi:schemaLocation="http://doctrine-project.org/schemas/orm/doctrine-mapping http://doctrine-project.org/schemas/orm/doctrine-mapping.xsd">
+
+    <mapped-superclass name="Setono\SyliusRestockNotificationPlugin\Entity\StockNotification" table="setono_sylius_restock_notification">
```
Comment (loevgaard) [inline src/Resources/config/doctrine/StockNotification.orm.xml]:
```suggestion
    <mapped-superclass name="Setono\SyliusRestockNotificationPlugin\Entity\StockNotification" table="setono_sylius_restock__notification">
```

It is a convention for my plugins that I have `[namespace]__[name of table]` for table names

File: src/Resources/config/doctrine/StockNotification.orm.xml [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+        <id name="id" type="integer">
+            <generator strategy="AUTO" />
+        </id>
+
+        <field name="email" type="string" nullable="false" />
```
Comment (loevgaard) [inline src/Resources/config/doctrine/StockNotification.orm.xml]:
```suggestion
        <field name="email" type="string"/>
```

File: src/Resources/config/doctrine/StockNotification.orm.xml [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+        <field name="locale" type="string" nullable="false" />
```
Comment (loevgaard) [inline src/Resources/config/doctrine/StockNotification.orm.xml]:
```suggestion
        <field name="locale" type="string"/>
```

File: src/Resources/config/doctrine/StockNotification.orm.xml [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+
+        <many-to-one field="product" target-entity="Sylius\Component\Product\Model\ProductInterface">
+            <join-column on-delete="CASCADE" />
```
Comment (loevgaard) [inline src/Resources/config/doctrine/StockNotification.orm.xml]: Remember to name the column when you use it in an index

File: src/Resources/config/services.yml [hunk 19]
```diff
@@ -0,0 +1,47 @@
+services:
```
Comment (loevgaard) [inline src/Resources/config/services.yml]: Configure services in XML

File: src/Resources/config/services.yml [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+    setono_restock_notification.shop.layout.js:
+        class: Sylius\Bundle\UiBundle\Block\BlockEventListener
+        arguments: ['@@SetonoSyliusRestockNotificationPlugin/Shop/_javascripts.html.twig']
+        tags:
+            - { name: kernel.event_listener, event: sonata.block.event.sylius.shop.layout.javascripts, method: 'onBlockEvent'}
+
+    setono_restock_notification.shop.layout.css:
+        class: Sylius\Bundle\UiBundle\Block\BlockEventListener
+        arguments: ['@@SetonoSyliusRestockNotificationPlugin/Shop/_stylesheets.html.twig']
```
Comment (loevgaard) [inline src/Resources/config/services.yml]: It says stylesheets here, but you listen to a javascript event or something?

File: src/Resources/public/js/shop/custom.js [hunk 21]
```diff
@@ -0,0 +1,31 @@
+$(document).on('submit', '#setono-rn--form', function () {
```
Comment (loevgaard) [inline src/Resources/public/js/shop/custom.js]: Without know I suppose you are missing a document ready here?

File: src/Resources/views/emails/stock_notification.html.twig [hunk 22]
```diff
@@ -0,0 +1,9 @@
+{# @var product \Sylius\Component\Core\Model\ProductInterface #}
+
+{% block subject %}
+    Product {{ product.name }} has been restocked.
+{% endblock %}
+
+{% block body %}
+    Product {{ product.name }} has been restocked.
```
Comment (loevgaard) [inline src/Resources/views/emails/stock_notification.html.twig]: There should be an URL here

File: tests/Application/.env [hunk 23]
```diff
@@ -0,0 +1,23 @@
...
+APP_ENV=dev
+APP_DEBUG=1
+APP_SECRET=EDITME
+###< symfony/framework-bundle ###
+
+###> doctrine/doctrine-bundle ###
+# Format described at http://docs.doctrine-project.org/projects/doctrine-dbal/en/latest/reference/configuration.html#connecting-using-a-url
+# For a sqlite database, use: "sqlite:///%kernel.project_dir%/var/data.db"
+# Set "serverVersion" to your server version to avoid edge-case exceptions and extra database calls
+DATABASE_URL=mysql://root:root@127.0.0.1/sylius_restock_plugin_%kernel.environment%?serverVersion=5.5
```
Comment (loevgaard) [inline tests/Application/.env]:
```suggestion
DATABASE_URL=mysql://root:root@127.0.0.1/setono_sylius_restock_notification_%kernel.environment%?serverVersion=5.5
```

File: tests/Application/.env.test [hunk 24]
```diff
@@ -0,0 +1,3 @@
+APP_SECRET='ch4mb3r0f5ecr3ts'
+
+KERNEL_CLASS='Tests\Acme\SyliusExamplePlugin\Application\Kernel'
```
Comment (loevgaard) [inline tests/Application/.env.test]: Fix namespace

File: tests/Behat/Resources/services.xml [hunk 25]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <defaults public="true"/>
+        <!-- Put your Behat services here -->
+
+        <service id="acme_sylius_example.context.ui.shop.welcome"
```
Comment (loevgaard) [inline tests/Behat/Resources/services.xml]: Just remove these services

File: src/Resources/public/js/shop/custom.js (hunk 21 above)
Comment (zmitic) [reply src/Resources/public/js/shop/custom.js]: No; ``$(document).ready`` is triggered only on full page reload. But by using above syntax, it will be triggered even if form is created via ajax or plain JS.

File: src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php [hunk 26]
```diff
@@ -0,0 +1,35 @@
...
+class SetonoSyliusRestockNotificationExtension extends ConfigurableExtension implements PrependExtensionInterface
+{
+    protected function loadInternal(array $mergedConfig, ContainerBuilder $container): void
+    {
+    }
+
+    /**
+     * Prepend twig paths for overriding templates
+     */
+    public function prepend(ContainerBuilder $container): void
```
[the remark this answers is not in the document]
Comment (zmitic) [reply src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php]: https://symfony.com/doc/4.1/bundles/prepend_extension.html

I used it to avoid manual copying of templates or using symlinks. Check other bundles, including yours; they require manual copying.

With above code, there is no need for that.

File: src/Form/Validator/CartItemAvailability.php (hunk 5 above)
Comment (zmitic) [reply src/Form/Validator/CartItemAvailability.php]: This and above are WIP; serve to validate variant availability.

Comment (loevgaard) [reply src/Form/Validator/CartItemAvailability.php]: Validations should go in xml files

File: src/Resources/public/js/shop/custom.js (hunk 21 above)
Comment (loevgaard) [reply src/Resources/public/js/shop/custom.js]: Usually I would use a syntax like:

```javascript
$(function() {
    $('#setono-rn--form').on('submit', function() {})
});
```

But I see your point. The event bubbles up all the way to the `document` eventually

File: src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php (hunk 26 above)
[the remark this answers is not in the document]
Comment (loevgaard) [reply src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php]: I understand that, but what happens if the user already has overridden a template? Or several plugins require the user to copy things?

**Comment by loevgaard** [comment]
These checks should also be free from errors:

`vendor/bin/phpstan analyse -c phpstan.neon -l max src/`
`vendor/bin/ecs check src/`

File: src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php (hunk 26 above)
[the remark this answers is not in the document]
Comment (zmitic) [reply src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php]: If that happens, all will be displayed if they use {% extends .... %} syntax (which they should).

Copying files doesn't change anything except adding complexity; it would still work as my syntax.

[the remark this answers is not in the document]
Comment (zmitic) [reply src/DependencyInjection/SetonoSyliusRestockNotificationExtension.php]: Also, after this plugin is done I plan to create PR for this to be implemented in core; it is super useful. 
But I will first ask; maybe it already is but noone documented the usage.
